3scale/porta
# THREESCALE-10233: Support OpenAPI 3.1 for ActiveDocs

**What this PR does / why we need it**:

Add support for OpenAPI 3.1:
- upgrade to a newer `swagger-ui` version
- use a different validation library for validating OpenAPI - [json_schemer](https://github.com/davishmcclurg/json_schemer)

The new library is required for 3.1 validation, because it's schema depends on JSON Schema version `2020-12`, which is not supported by the current library [json-schema](https://github.com/voxpupuli/json-schema).

We actually use a fork of this library to support `draft7` of JSON Schema: https://github.com/voxpupuli/json-schema/compare/master...3scale:json-schema:master This was required for policies validation. And we eventually might want to switch to `json_schemer` for policies validation.

However, I found out that validation of Swagger 1.2 does not work out of the box with `json_schemer`, so I kept the old validation for it. We don't really "support" Swagger 1.2 anymore, but it kind of still works, so probably we shouldn't break it.

And for Swagger 2.0 I think `json_schemer` could be used, but I decided to keep the current solution.

On the other hand, I decided to move OpenAPI 3.0 to new validation :grimacing:  Just because it's supported out of the box by `json_schemer`, without providing the JSON spec for it.
The drawback is that the validation error messages change in case there are some issues with the spec, but I think it's not a big deal.

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-10233

**Verification steps** 

Create an API spec following OpenAPI v3.1 and make sure it's rendered properly. Example minimal spec:
```
{
  "openapi": "3.1.0",
  "info": {
    "title": "Sample API",
    "description": "Sample API description",
    "version": "1.0.0"
  },
  "servers": [
    {
      "url": "https://echo-api.3scale.net",
      "description": "Echo API"
    }
  ],
  "paths": {
    "/hello": {
      "get": {
        "operationId": "hello",
        "description": "Hello",
        "tags": [
          "greeting"
        ],
        "responses": {
          "200": {
            "description": "OK"
          },
          "400": {
            "description": "Bad request"
          }
        }
      }
    }
  }
}
```

**Special notes for your reviewer**:

There is some strange behavior in the latest versions of the library, apparently caused by this PR: https://github.com/swagger-api/swagger-ui/pull/9762 which first appears in the version [v5.13.0](https://github.com/swagger-api/swagger-ui/releases/tag/v5.13.0). This results in all fields in some requests to be auto-filled with `string` value, which is kind of considered a "default" or "example" value. This is obviously unexpected and very annoying.


![swagger-ui-string-initial-value](https://github.com/user-attachments/assets/42197b08-5dc6-43e1-9359-f14c11562907)


One possible way of solving it is to define the default values for all fields, but I decided to downgrade to an older version that doesn't have this problem. The main reason being that I consider this behavior wrong and I think it will need to be fixed in `swagger-ui`. I will open an issue for it, but I am not very optimistic, as there are some issues which are not quite the same, but related, and they seem to be getting little attention...e.g. https://github.com/swagger-api/swagger-ui/issues/5776
